File: oracle/tax_calculator.test.js
import { describe, expect, it } from "vitest";

import { computeTax } from "./tax_calculator.js";

describe("Test computeTax", () => {
  const mockTable = [
    { upTo: 100, rate: 0.0 },
    { upTo: 200, rate: 0.1 },
    { upTo: 300, rate: 0.2 },
    { upTo: Infinity, rate: 0.05, type: "flat" },
  ];

  it("should return 0 when the total value is 0 (avoiding NaN)", () => {
    const result = computeTax(0, 0, mockTable);
    expect(result.amount).toBe(0);
    expect(result.marginalRate).toBe(mockTable[0].rate);
    expect(result.deduction).toBe(0);
  });

  it("should return 0 for values in the first bracket (0%)", () => {
    const result = computeTax(50, 50, mockTable);
    expect(result.amount).toBe(0);
    expect(result.marginalRate).toBe(0);
  });

  it("should calculate progressive tax correctly across two brackets", () => {
    const result = computeTax(150, 150, mockTable);
    expect(result.amount).toBe(5);
    expect(result.marginalRate).toBe(0.1);
  });

  it("should calculate progressive tax correctly across three brackets", () => {
    const result = computeTax(250, 250, mockTable);
    expect(result.amount).toBe(20);
    expect(result.marginalRate).toBe(0.2);
  });

  it("should jump to flat rate logic when the threshold is hit", () => {
    const result = computeTax(400, 400, mockTable);
    expect(result.amount).toBe(20);
    expect(result.marginalRate).toBe(0.05);
  });

  it("should apply proportionality correctly for partial shares", () => {
    const result = computeTax(125, 250, mockTable);
    expect(result.amount).toBe(10);
  });
});

File: src/test/kotlin/imt/TaxCalculatorTest.kt
package imt

import kotlin.test.Test
import kotlin.test.assertEquals

/** Ported from oracle/tax_calculator.test.js. */
class TaxCalculatorTest {

    private val mockTable = listOf(
        Bracket(upTo = 100.0, rate = 0.0),
        Bracket(upTo = 200.0, rate = 0.1),
        Bracket(upTo = 300.0, rate = 0.2),
        Bracket(upTo = Double.POSITIVE_INFINITY, rate = 0.05, isFlat = true),
    )

    @Test
    fun `returns 0 when the total value is 0 (avoiding NaN)`() {
        val result = computeTax(0.0, 0.0, mockTable)
        assertEquals(0.0, result.amount)
        assertEquals(mockTable[0].rate, result.marginalRate)
        assertEquals(0.0, result.deduction)
    }

    @Test
    fun `returns 0 for values in the first bracket (0 percent)`() {
        val result = computeTax(50.0, 50.0, mockTable)
        assertEquals(0.0, result.amount)
        assertEquals(0.0, result.marginalRate)
    }

    @Test
    fun `calculates progressive tax correctly across two brackets`() {
        val result = computeTax(150.0, 150.0, mockTable)
        assertEquals(5.0, result.amount)
        assertEquals(0.1, result.marginalRate)
    }

    @Test
    fun `calculates progressive tax correctly across three brackets`() {
        val result = computeTax(250.0, 250.0, mockTable)
        assertEquals(20.0, result.amount)
        assertEquals(0.2, result.marginalRate)
    }

    @Test
    fun `jumps to flat rate logic when the threshold is hit`() {
        val result = computeTax(400.0, 400.0, mockTable)
        assertEquals(20.0, result.amount)
        assertEquals(0.05, result.marginalRate)
    }

    @Test
    fun `applies proportionality correctly for partial shares`() {
        val result = computeTax(125.0, 250.0, mockTable)
        assertEquals(10.0, result.amount)
    }
}
